Firestore/Source/Local/FSTLevelDBQueryCache.mm

```diff
@@ -148,6 +163,53 @@ - (void)setLastRemoteSnapshotVersion:(SnapshotVersion)snapshotVersion {
   _db.currentTransaction->Put([FSTLevelDBTargetGlobalKey key], self.metadata);
 }
 
+- (void)enumerateTargetsUsingBlock:(void (^)(FSTQueryData *queryData, BOOL *stop))block {
+  // Enumerate all targets, give their sequence numbers.
+  std::string targetPrefix = [FSTLevelDBTargetKey keyPrefix];
+  auto it = _db.currentTransaction->NewIterator();
+  it->Seek(targetPrefix);
+  BOOL stop = NO;
+  for (; !stop && it->Valid() && absl::StartsWith(it->key(), targetPrefix); it->Next()) {
+    FSTQueryData *target = [self decodedTarget:it->value()];
+    block(target, &stop);
+  }
+}
+
+- (void)enumerateOrphanedDocumentsUsingBlock:
+    (void (^)(FSTDocumentKey *docKey, FSTListenSequenceNumber sequenceNumber, BOOL *stop))block {
+  std::string documentTargetPrefix = [FSTLevelDBDocumentTargetKey keyPrefix];
+  auto it = _db.currentTransaction->NewIterator();
+  it->Seek(documentTargetPrefix);
+  FSTListenSequenceNumber nextToReport = 0;
+  FSTDocumentKey *keyToReport = nil;
+  FSTLevelDBDocumentTargetKey *key = [[FSTLevelDBDocumentTargetKey alloc] init];
+  BOOL stop = NO;
+  for (; !stop && it->Valid() && absl::StartsWith(it->key(), documentTargetPrefix); it->Next()) {
+    [key decodeKey:it->key()];
+    if (key.isSentinel) {
+      // if nextToReport is non-zero, report it, this is a new key so the last one
+      // must be orphaned.
+      if (nextToReport != 0) {
```

Comment:
What exactly constitutes an orphaned document? Is it a document for which there exists only a sentinel row?

What if the first two rows contain sentinels?
  * First iteration, we'll see
    * key.isSentinel, nextToReport == 0, so no report
    * set nextToReport == 0
  * Second iteration, we'll see
    * key.isSentental, nextToReport == 0, so no report

It's possible I'm misunderstanding so an implementation comment describing which rows should be reported is probably warranted regardless.

I can't quite tell, but are we missing test coverage over this and similar cases?

Reply:
On the first iteration, we would not set `nextToReport` to `0`. We would set it to the sequence number of the sentinel row, which is non-zero. On the next iteration, we would see the sentinel key and have a non-zero `nextToReport`, so the first sentinel would get reported. On the third iteration, if we have another sentinel key, the second would get reported. If the third row is not a sentinel, it must be a valid target for the second key, so it should not get reported.

Reply:
Clarified `orphaned` as `not a member of any targets`.